core/include/gnuradio-4.0/BlockModel.hpp

```diff
@@ -546,6 +546,186 @@ class BlockWrapper : public BlockModel {
+    result["min_buffer_size"s] = edge.minBufferSize();
+    result["edge_name"s]       = std::string(edge.name());
+
+    result["buffer_size"s] = edge.bufferSize();
+    result["edge_state"s]  = std::string(magic_enum::enum_name(edge.state()));
+    result["n_readers"s]   = edge.nReaders();
+    result["n_writers"s]   = edge.nWriters();
+    result["type"s]        = std::string(magic_enum::enum_name(edge.edgeType()));
+
+    return result;
+};
+
+namespace BlockSerializationFlags {
+enum Flags : int {
+    Data       = 1,  //
+    StaticData = 2,  //
+    Ports      = 4,  //
+    Settings   = 8,  //
+    Children   = 16, //
+    All        = 0xFF
+};
+}
+
+template<typename PluginLoader>
+requires(not std::is_pointer_v<std::remove_cvref_t<PluginLoader>>)
+inline property_map serializeBlock(PluginLoader& pluginLoader, const std::shared_ptr<BlockModel>& block, int flags) {
+    using namespace std::string_literals;
+
+    auto serializePortOrCollection = [](const auto& portOrCollection) {
+        // clang-format off
+            // TODO: Type names can be mangled. We need proper type names...
+            return std::visit(meta::overloaded{
+                    [](const gr::DynamicPort& port) {
+                        return property_map{
+                            {"name"s, std::string(port.name)},
+                            {"type"s, port.typeName()}
+                        };
+                    },
+                    [](const BlockModel::NamedPortCollection& namedCollection) {
+                        return property_map{
+                            {"name"s, std::string(namedCollection.name)},
+                            {"size"s, namedCollection.ports.size()},
+                            {"type"s, namedCollection.ports.empty() ? std::string() : std::string(namedCollection.ports[0].typeName()) }
+                        };
+                    }},
+                portOrCollection);
+        // clang-format on
+    };
+
+    property_map result;
+    result.emplace("id", pluginLoader.registry().typeName(block));
+    result.emplace("unique_name"s, std::string(block->uniqueName()));
+
+    if (flags & BlockSerializationFlags::Data) {
+        result.emplace("name"s, std::string(block->name()));
+        result.emplace("ui_constraints"s, std::get<property_map>(block->settings().get("ui_constraints").value()));
+    }
+
+    if (flags & BlockSerializationFlags::StaticData) {
+        result.emplace("type_name"s, std::string(block->typeName()));
+        result.emplace("is_blocking"s, block->isBlocking());
+        result.emplace("block_category"s, std::string(magic_enum::enum_name(block->blockCategory())));
+        result.emplace("ui_category"s, std::string(magic_enum::enum_name(block->uiCategory())));
+        result.emplace("meta_information"s, block->metaInformation());
+    }
+
+    if (flags & BlockSerializationFlags::Settings) {
+        result.emplace("settings"s, block->settings().getStored().value_or(property_map()));
```

Comment:
This looks like a lot like hardcoded magic strings. We have a compile-time type introspection system. Why not use that?

This is very brittle in case names are changed in the `Block<T>` definition.

Reply:
Feels also a bit like a PMO: what is this selectivity trying to achieve?

Reply:
Does this need to be part of the BlockModel?

Reply:
On the contrary,  if the internal names change, the format will not and pre-rename GRC files will remain valid. Which would not be the case if this was based on the reflection data.

Reply:
I am OK with (optionally) suppressing the meta_information (since this is needed only once per block). The other information (block settings, a few limit UI constraints) is nearly always needed in full. Maybe some caching would help there.

We follow the principle of 'the code is the single source of truth'.... thus also the reliance on compile-time reflection since this avoids a lot (!!) of development errors. 

Please do not hard-code string literals in the main code. For unit-tests it's less critical but the library should use only proper names and if possible those derived from the class design... thus also the naming conventions.